ДЗ по спринту 11

ovseychik opened the pull request.

Commit:
Sprint 11 homework.

Commit:
Sprint 11 homework.

faritowich commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -84,8 +85,90 @@
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginHorizontal="8dp"
+        android:layout_marginTop="113dp"
```
Лучше не центрировать огромным марждином, на разных устройствах будет выглядеть по-разному. Лучше как-то отрефакторить верстку, например блок с ошибкой поместить посередине FrameLayout'а, который будет во весь экран

ovseychik replied:
Поправил верстку, спасибо!

faritowich commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -84,8 +85,90 @@
+        android:gravity="center_horizontal"
+        android:orientation="vertical"
+        android:visibility="gone"
+        tools:visibility="visible">
```
Лучше реализовать показ заглушек не разными лейаутами placeHolderNoInternet и placeHolderNothingFound, а объединить в один, а изображение, текст и видимость кнопки "обновить" задавать в активити. Можно создать в активити отдельный метод, в аргументы которого передается энам статуса поиска (этот энам надо создать и сделать в нем статусы не найдено, нет сети, успех) и в зависимости от переданного энама задавать картинку, сеттить  текст ошибки, видимость кнопки Обновить и т,д.. Тогда будет меньше загружен xml-файл, а также упростится логика работы в активити

ovseychik replied:
Поправил. Логику завязал на enum. Верстка скорректирована, теперь картинка и сообщения передаются в зависимости от ошибки.

faritowich commented on app/src/main/java/com/example/playlistmaker/TrackResponse.kt:
```diff
@@ -0,0 +1,6 @@
+
+import java.util.ArrayList
+
+class TrackResponse(val results: ArrayList<Track>) {
```
Модельные классы лучше объявлять как data class

ovseychik replied:
Поправил, спасибо!

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,94 +1,118 @@
         }
 
+        // инициализируем recyclerview
+        val trackRecycler = findViewById<RecyclerView>(R.id.trackList)
```
Раз уж используется lateinit, лучше все переменные, отвечающие за View, объявлять с lateinit

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,94 +1,118 @@
         }
     }
 
+    fun showPlaceholder(text: String) {
```
Если метод/переменная используется только в классе, лучше делать их приватными

ovseychik replied:
Поправил, спасибо!

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,94 +1,118 @@
+            TRACK_NOT_FOUND -> placeholderNoResult.visibility = View.VISIBLE
+            else -> placeholderNoNetwork.visibility = View.VISIBLE
+        }
+    }
```
Идея хорошая, но лучше завязаться не на строковые константы, а сделать под это дело enum класс статуса поиска и в зависимости от энама показывать тот и или иной результат, показывать или скрывать какие-то элементы.

ovseychik replied:
Попробовал завязаться на enum класс.

Commit:
Спринт 11. Исправлены критические замечания от ревьюера:
- Корректно отрабатывает плейсхолдер "Нет сети".
- Кнопка очистки поисковой строки очищает экран.

Commit:
Спринт 11. Учтены дополнительные рекомендации ревьюера:
- Перерисованы плейсхолдеры: он теперь один, картинка и текст передаются при обработке ошибки. Логика завязана на ENUM.
- Скорректированы ресурсы с картинками, чтобы подставлялись корректные в зависимости от темы День/Ночь.
- Класс TrackResponse.kt теперь data-класс.

ovseychik commented:
Учтены дополнительные рекомендации ревьюера. Одна проблема: сейчас android studio ругается на картинки в плейсхолдерах. 
```
Error:(1, 2) The drawable "nothing_found" in drawable-night has no declaration in the base `drawable` folder or in a `drawable-`*density*`dpi` folder; this can lead to crashes when the drawable is queried in a configuration that does not match this qualifier

Error:(1, 2) The drawable "no_internet" in drawable-night has no declaration in the base `drawable` folder or in a `drawable-`*density*`dpi` folder; this can lead to crashes when the drawable is queried in a configuration that does not match this qualifier
```

Насколько это критично, и требуется ли иправить? Приложение компилируется, в эмуляторе запускается.
![image](https://user-images.githubusercontent.com/5370490/233853614-cfd59f24-e840-4f02-860b-a97045861b25.png)

faritowich commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -1,94 +1,131 @@
+        //SUCCESS("OK 200"),
+        CHECK_NETWORK("CHECK NETWORK"),
+        NOTHING_FOUND("NO RESULT")
+    }
```
Тут можно обойтись без message, он не используется
